dali/operators/generic/histogram/histogram.cc

```
@@ -0,0 +1,685 @@
+    }
+  }
+  return ranges;
+}
+
+template <typename Ty_>
+struct CVMatType {
+  static int get(int) {
+    DALI_ENFORCE(false, "Unreachable - invalid type");
+  }
+};
+
+template <>
+struct CVMatType<uint8_t> {
+  static int get(int nchannel) noexcept {
+    return CV_MAKETYPE(CV_8U, nchannel);
+  }
+};
+
+template <>
+struct CVMatType<uint16_t> {
+  static int get(int nchannel) noexcept {
+    return CV_MAKETYPE(CV_16U, nchannel);
+  }
+};
+
+template <>
+struct CVMatType<float> {
+  static int get(int nchannel) noexcept {
+    return CV_MAKETYPE(CV_32F, nchannel);
+  }
+};
+
+template <>
+struct CVMatType<double> {
+  static int get(int nchannel) noexcept {
+    return CV_MAKETYPE(CV_64F, nchannel);
+  }
+};
+
+template <typename Type, typename ScratchAlloc, typename Coll>
+TensorListView<StorageCPU, const Type> transpose_view(
+    dali::ThreadPool &thread_pool, ScratchAlloc &scratch,
+    const TensorListView<StorageCPU, const Type> &in_view, const Coll &transpose_axes_order) {
+  const auto &in_shapes = in_view.shape;
+
+  TensorListShape<> transposed_shapes;
+  permute_dims(transposed_shapes, in_shapes, transpose_axes_order);
+  std::vector<Type *> tmp_pointers;
+  tmp_pointers.reserve(transposed_shapes.num_samples());
+
+  for (int i = 0; i < transposed_shapes.num_samples(); ++i) {
+    auto tmp = scratch.template AllocTensor<mm::memory_kind::host, Type>(transposed_shapes[i]);
+    tmp_pointers.push_back(tmp.data);
+  }
+
+  TensorListView<StorageCPU, Type> transpose_out_view(std::move(tmp_pointers),
+                                                      std::move(transposed_shapes));
+
+  for (int i = 0; i < transpose_out_view.num_samples(); ++i) {
+    thread_pool.AddWork([&, i](int thread_id) {
+      auto perm = make_span(transpose_axes_order);
+      kernels::Transpose(transpose_out_view[i], in_view[i], perm);
+    });
+  }
+  thread_pool.RunAll(true);
+  return reinterpret<const Type>(transpose_out_view, transpose_out_view.shape);
```

```suggestion
  return transpose_out_view;
```
would do, I think

---

Yeah, the conversion to const view is implicit. Reinterpret is a heavy hammer - like with a cast, it will hide any possible errors until run-time.